### ISSUE TITLE
"no-unused-state" w/ Used State as prevState

### ISSUE THREAD

**Issue by alistair-hmh** [yannickcr/eslint-plugin-react#2209]
Version: 7.12.4

Code:

```js
import React from 'react';

class ComponentX extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      counter: 0,
      // Unused state field: 'allowUpdates'eslint(react/no-unused-state)
      allowUpdates: true
    };
  }

  update() {
    this.setState(prevState => {
      if (prevState.allowUpdates) {
        // (... but the state is used here?)
        const nextCounter = prevState.counter + 1;

        return {
          counter: nextCounter
        };
      }
    });
  }

  render() {
    return (
      <div>
        <div>Counter = {this.state.counter}</div>{" "}
        <button onClick={() => this.update()}>Update</button>
      </div>
    );
  }
}

export default ComponentX;
```

CodeSandbox: https://codesandbox.io/s/2o9pvqnyky

Screengrab:

![Screen Shot 2019-03-20 at 3 27 41 PM](https://user-images.githubusercontent.com/38665769/54713330-bb9c4f80-4b24-11e9-8dd6-76ede5494085.png)

**Comment by ljharb** [comment]
If you instead do this, does it still complain?
```js
update() {
    this.setState(({ counter, allowUpdates }) => {
      if (allowUpdates) {
        // (... but the state is used here?)
        const nextCounter = counter + 1;

        return {
          counter: nextCounter
        };
      }
    });
  }
```

**Comment by alistair-hmh** [comment]
Yes, unfortunately, it still complains when destructuring like that.

**Comment by jzabala** [comment]
This seems to be working find with versions:

```
"eslint": "^7.3.1",
"eslint-plugin-react": "^7.20.2"
```

![Screen Shot 2020-07-08 at 3 21 23 PM](https://user-images.githubusercontent.com/1315054/86961268-ed2c9600-c12e-11ea-967a-1db0b27b0e0f.png)
